modules/SerialCommHub/tiny_modbus_rtu.cpp

```diff
@@ -103,92 +131,113 @@ static std::vector<uint16_t> decode_reply(const uint8_t* buf, int len, uint8_t e
+            throw ModbusException("Modbus exception: NACK");
             break;
         case 0x08:
-            EVLOG_error << "Modbus exception: Memory parity error";
+            throw ModbusException("Modbus exception: Memory parity error");
             break;
         case 0x09:
-            EVLOG_error << "Modbus exception: Out of resources";
+            throw ModbusException("Modbus exception: Out of resources");
             break;
         case 0x0A:
-            EVLOG_error << "Modbus exception: Gateway path unavailable";
+            throw ModbusException("Modbus exception: Gateway path unavailable");
             break;
         case 0x0B:
-            EVLOG_error << "Modbus exception: Gateway target device failed to respond";
+            throw ModbusException("Modbus exception: Gateway target device failed to respond");
             break;
         default:
-            EVLOG_error << "Modbus exception: Unknown";
+            throw ModbusException("Modbus exception: Unknown");
         }
-        return result;
     }
+
+    // For a write reply we always get 4 bytes
+    uint8_t byte_cnt = 4;
+    int start_of_result = RES_TX_START_OF_PAYLOAD;
+    bool even_byte_cnt_expected = false;
+
+    // Was it a read reply?
+    switch (function) {
+    case FunctionCode::WRITE_SINGLE_COIL:
+    case FunctionCode::WRITE_SINGLE_HOLDING_REGISTER:
+    case FunctionCode::WRITE_MULTIPLE_COILS:
+    case FunctionCode::WRITE_MULTIPLE_HOLDING_REGISTERS:
+        // no - nothing to do
+        break;
+    case FunctionCode::READ_MULTIPLE_HOLDING_REGISTERS:
+    case FunctionCode::READ_INPUT_REGISTERS:
+        // yes - for 16-bit wide registers thus we can assume a even byte count
+        even_byte_cnt_expected = true;
+        [[fallthrough]];
+    case FunctionCode::READ_COILS:
+    case FunctionCode::READ_DISCRETE_INPUTS:
+        // yes
+        // adapt byte count and starting pos
+        byte_cnt = buf[RES_RX_LEN_POS];
+        start_of_result = RES_RX_START_OF_PAYLOAD;
+        break;
+    default:
+        throw UnimplementedFunction("Missing implementation for function code " +
+                                    FunctionCode_to_string_with_hex(function));
+    }
+
+    // check if result is completely in received data
+    if (start_of_result + byte_cnt > len) {
+        throw IncompletePacketException("Result data not completely in received message.");
+    }
+
+    // check if result is completely in received data
+    if (even_byte_cnt_expected && byte_cnt % 2 == 1) {
+        throw OddByteCountException("For " + FunctionCode_to_string_with_hex(function) +
+                                    " an even byte count is expected in the response.");
+    }
+
+    // ready to copy actual result data to output, so pre-allocate enough memory for the output
+    result.reserve((byte_cnt + 1) / 2);
+
+    for (int i = start_of_result; i < start_of_result + byte_cnt; i += 2) {
+        uint16_t t;
+        memcpy(&t, buf + i, 2);
```

Comment:
wont this read invalid memory in byte_cnt is odd?

Reply:
Some lines above, it is ensured that `byte_cnt` is even.

Reply:
The last item in the buffer it writes to potentially is buf[start_of_result + byte_cnt] for odd byte count (and no expected even byte count).

As far as I can see the

if (start_of_result + byte_cnt > len)

does not catch it. So I also think it writes one byte past the buffer.

Reply:
you also don't need to copy element-wise -  just doing memcpy on the whole buffer or constructing the result directly from the buffer should work (given that you handle the odd case)

Reply:
> The last item in the buffer it writes to potentially is buf[start_of_result + byte_cnt] for odd byte count (and no expected even byte count).

I re-reviewed the code. (This is part of the original implementation, BTW, not of the changes in this PR.)
Indeed, when receiving an odd number of bytes, the `memcpy()` will read beyond the end of `buf`.

E.g. read coils, five bytes:
```C++
    for (int i = start_of_result; i < start_of_result + byte_cnt; i += 2) {
        uint16_t t = 0;
        memcpy(&t, buf + i, 2);
        t = be16toh(t);
        result.push_back(t);
    }
```
len = 8
start_of_result = 3
byte_cnt = 5
Then `i` will take the values 3, 5, 7. The last iteration will `memcopy()` two bytes from `buf[7]`, the second of which is beyond its end. This is either undefined behavior or illegal access (likely not due to alignment), and potentially returns data not belonging to the response.

I don't know if there is even a point of re-interpreting the return value of "read coils" and "read inputs" as big-endian (I understand these are bit fields). I am open for comments. I will just keep the be16 conversion in place, and make sure `memcpy()` places the single, final byte from `buf` into the correct position.

Is that okay for you?

Reply:
> you also don't need to copy element-wise - just doing memcpy on the whole buffer or constructing the result directly from the buffer should work (given that you handle the odd case)

You cannot just memcopy from the byte array to the word array, as the bytes need to (potentially) be converted from Modbus's big-endian order. (You could do some casting and union tricks for conversion, but for me, `memcpy()` is good to read.)